src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -46,31 +46,51 @@ public IModelBinder CreateBinder(ModelBinderFactoryContext context)
                 throw new ArgumentNullException(nameof(context));
             }
 
-            // We perform caching in CreateBinder (not in CreateBinderCore) because we only want to
-            // cache the top-level binder.
             IModelBinder binder;
-            if (context.CacheToken != null && _cache.TryGetValue(context.CacheToken, out binder))
+            if (TryGetCachedBinder(context.Metadata, context.CacheToken, out binder))
             {
                 return binder;
             }
-
+            
+            // Perf: We're calling the Uncached version of the API here so we can:
+            // 1. avoid allocating a context when the value is already cached
+            // 2. avoid checking the cache twice when the value is not cached
             var providerContext = new DefaultModelBinderProviderContext(this, context);
-            binder = CreateBinderCore(providerContext, context.CacheToken);
+            binder = CreateBinderCoreUncached(providerContext, context.CacheToken);
             if (binder == null)
             {
                 var message = Resources.FormatCouldNotCreateIModelBinder(providerContext.Metadata.ModelType);
                 throw new InvalidOperationException(message);
             }
 
-            if (context.CacheToken != null)
+            AddToCache(context.Metadata, context.CacheToken, binder);
```

Comment:
If I'm reading this correctly, it's not possible for `binder` to be a `PlaceholderBinder` here. Might be worth an assertion. If I'm wrong, suggest adding `placeholder.Inner` (which _must_ not be `null`) to the cache instead.